File: examples/python/ds/data/publisher.py
from ds.base import *
from ds.model.publisher import *
from ds.model._dataset import Dataset
from quartz.data.table import Table


class PublisherData(Dataset.Publishers):

    def __init__(self, db):
        self._table = PublisherTable("publisher", db)

    def add(self, x, debug:bool=False):
        if isinstance(x, Publisher) and (x.id is None):
            x.id = self._table.insert(x, debug)
            return x
        elif self._replace(x, debug):
            return x
        self._table.insert(x, debug)
        return x

    def remove(self, x, debug:bool=False):
        return self._table.delete(x, debug)

    def push(self, x, field:Fields=None, debug:bool=False):
        if self._replace(x, True, debug):
            return x
        if isinstance(x, Publisher) and isinstance(field, Fields):
            match field:
                case Fields.Parent:
                    x = Table.Info("publisher", [ ("id", x.id), ("parent_id", x.parent_id) ], 1)
                case Fields.Name:
                    x = Table.Info("publisher", [ ("id", x.id), ("name", x.name) ], 1)
                case _:
                    x = None
        return self._table.update(x, debug) if (x is not None) else None

    def pull(self, x:Publisher, field:Fields=None, debug:bool=False):
        match field:
            case None:
                row = self._table.one(Table.Info("publisher", [ ("id", x.id), ("parent_id", None), ("name", None) ], 1), debug=debug)
                if (isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0):
                    x.parent_id = row[0]
                    x.name = row[1]
            case Fields.Parent:
                row = self._table.one(Table.Info("publisher", [ ("id", x.id), ("parent_id", None) ], 1), debug=debug)
                if (isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0):
                    x.parent_id = row[0]
            case Fields.Name:
                row = self._table.one(Table.Info("publisher", [ ("id", x.id), ("name", None) ], 1), debug=debug)
                if (isinstance(row, tuple) or isinstance(row, list)) and (len(row) > 0):
                    x.name = row[0]
        return x

    def has(self, x, debug:bool=False) -> bool:
        return self.ref(x, debug) is not None

    def clear(self, debug:bool=False):
        self._table.deleteAll(debug)

    def refb(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.begin(x, 'pk', order, limit, debug)

    def refn(self, cursor) -> Reference:
        return self.next(cursor, self._reference)

    def ref(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.one(x, self._reference, 'pk', order, limit, debug)

    def refs(self, x=None, order:list[str]=None, limit:str=None, debug:bool=False):
        return self.all(x, self._reference, 'pk', order, limit, debug)

    def begin(self, x=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        return self._table.begin(x, columns, order, limit, debug)

    def next(self, cursor, builder=None):
        fn = self._create if (builder is None) else builder
        return self._table.next(cursor, fn)

    def one(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        fn = self._create if (builder is None) else builder
        return self._table.one(x, fn, columns, order, limit, debug)

    def all(self, x=None, builder=None, columns:str='full', order:list[str]=None, limit:str=None, debug:bool=False):
        fn = self._create if (builder is None) else builder
        return self._table.select(x, fn, columns, order, limit, debug)

    def _replace(self, x, allow_delete=False, debug:bool=False) -> bool:
        return False

    def _reference(self, x):
        return PublisherRef(id=x[0])

    def _create(self, x):
        return Publisher(id=x[0], parent_id=x[1], name=x[2])


class PublisherTable(Table):

    def _insert(self, x):
        # publisher
        if isinstance(x, Publisher):
            return Table.Info("publisher", [ x.id, x.parent_id, x.name ], 1, tag=("insert" if (x.id is None) else "insert_id"))

    def _update(self, x):
        # publisher
        if isinstance(x, Publisher):
            return Table.Info("publisher", [ x.id, x.parent_id, x.name ], 1, tag="update")

    def _delete(self, x):
        # "parent_publisher" publisher.parent_id → publisher.id (recursive)
        if isinstance(x, ParentPublisherRef):
            return Table.Info("publisher", [ ("parent", int(x)) ])
        # pk
        if isinstance(x, int) or isinstance(x, PublisherRef):
            return Table.Info("publisher", [ int(x) ], tag="pk")
        # name
        if isinstance(x, str):
            return Table.Info("publisher", [ x ], tag="name")

    def _select(self, x=None) -> tuple:
        # "parent_publisher" publisher.parent_id → publisher.id (recursive)
        if isinstance(x, ParentPublisherRef):
            return Table.Info("publisher", [ int(x) ], tag="parent")
        # pk
        if isinstance(x, int) or isinstance(x, PublisherRef):
            return Table.Info("publisher", [ int(x) ], tag="pk")
        # name (lookup)
        if isinstance(x, str):
            return Table.Info("publisher", [ x ], tag="name")
        # all
        if x is None:
            return Table.Info("publisher", tag="all")

File: examples/php/ds/data/publisher.php
<?php
namespace books;
require_once __DIR__ . '/../base.php';
require_once __DIR__ . '/../model/publisher.php';
require_once __DIR__ . '/../../common/data/base.php';
require_once __DIR__ . '/../../common/data/table.php';
require_once __DIR__ . '/../../common/error.php';

use quartz\Info, quartz\TagInfo, quartz\KeyInfo;
use quartz\Log;


class PublisherData {
    private \quartz\Table $table;

    function __construct(\quartz\Database $db) {
        $this->table = new PublisherTable("publisher", $db);
    }

    function add($x, bool $debug=false) {
        if (($x instanceof model\Publisher) && is_null($x->id)) {
            $x->id = $this->table->insert($x, $debug);
        }
        elseif (!$this->_replace($x, $debug)) {
            $this->table->insert($x, $debug);
        }
        return $x;
    }

    function push($x, bool $debug=false) {
        if (!$this->_replace($x, $debug)) {
            if (is_array($x) && (x[0] instanceof model\Publisher) && (x[1] instanceof Fields)) {
                // Individual fields
                [$o, $f] = $x;
                if ($f == Fields::Parent) {
                    $x = new KeyInfo("publisher", [ ["id", $o->id] ], [ ["parent_id", $o->parent_id] ]);
                }
                if ($f == Fields::Name) {
                    $x = new KeyInfo("publisher", [ ["id", $o->id] ], [ ["name", $o->name] ]);
                }
            }
            // Update
            return $this->table->update($x, $debug);
        }
    }

    function pull(model\Publisher $x, ?Fields $field=null, bool $debug=false) {
        if (is_null($field)) {
            $row = $this->table->select(KeyInfo("publisher", [ ["id", $x->id] ], [ ["parent_id", $x->parent_id], ["name", $x->name] ]), one:true, debug:$debug);
            $x->parent_id = $row[0];
            $x->name = $row[1];
        }
        elseif (field == Fields::Parent) {
            $x->parent_id = $this->table->select(KeyInfo("publisher", [ ["id", $x->id] ], [ ["parent_id", $x->parent_id] ]), one:true, debug:$debug)[0];
        }
        elseif (field == Fields::Name) {
            $x->name = $this->table->select(KeyInfo("publisher", [ ["id", $x->id] ], [ ["name", $x->name] ]), one:true, debug:$debug)[0];
        }
    }
    function remove($x, bool $debug=false) {
        return $this->table->delete($x, $debug);
    }

    function has($x, bool $debug=false): bool {
        return !is_null($this->ref($x, $debug));
    }

    function clear($debug=false) {
        $this->table->deleteAll(debug);
    }

    function refb($x=null, ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->begin($x, 'pk', $order, $limit, $debug);
    }

    function refn($cursor) {
        return $this->table->next($cursor, $this->reference(...));
    }

    function ref($x=null, ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->one($x, $this->reference(...), 'pk', $order, $limit, $debug);
    }

    function refs($x=null, ?array $order=null, ?string $limit=null, bool $debug=false): array {
        return $this->table->select($x, $this->reference(...), 'pk', $order, $limit, $debug);
    }

    function begin($x=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        return $this->table->begin($x, $columns, $order, $limit, $debug);
    }

    function next($cursor, ?callable $builder=null) {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->next($cursor, $fn);
    }

    function one($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false) {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->one($x, $fn, $columns, $order, $limit, $debug);
    }

    function all($x=null, ?callable $builder=null, ?string $columns='full', ?array $order=null, ?string $limit=null, bool $debug=false): array {
        $fn = is_null($builder) ? $this->create(...) : $builder;
        return $this->table->select($x, $fn, $columns, $order, $limit, $debug);
    }

    protected function _replace($x, bool $allow_delete=false, bool $debug=false) {
        return false;
    }

    protected function reference($x) {
        return new PublisherRef(id:$x[0]);
    }

    protected function create($x) {
        return new model\Publisher(id:$x[0], parent_id:$x[1], name:$x[2]);
    }
}


class PublisherTable extends \quartz\Table {

    function _insert($x): ?Info {
        // publisher
        if ($x instanceof model\Publisher) {
            return new Info("publisher", [ $x->id, $x->parent_id, $x->name ]);
        }
        return null;
    }

    function _update($x): ?Info {
        // publisher
        if ($x instanceof model\Publisher) {
            return new Info("publisher", [ $x->parent_id, $x->name, $x->id ]);
        }
        return null;
    }

    function _delete($x): ?Info {
        // "parent_publisher" publisher.parent_id → publisher.id (recursive)
        if ($x instanceof ParentPublisherRef) {
            return KeyInfo("publisher", [ ["parent", $x->id] ]);
        }
        // pk
        if (($x instanceof int) || ($x instanceof PublisherRef)) {
            return new TagInfo("publisher", "pk", [ $x->id ]);
        }
        // name
        if ($x instanceof str) {
            return new TagInfo("publisher", "name", [ x ]);
        }
        return null;
    }

    function _select($x=null): ?Info {
        // "parent_publisher" publisher.parent_id → publisher.id (recursive)
        if ($x instanceof ParentPublisherRef) {
            return new TagInfo("publisher", "parent", [ $x->id ]);
        }
        // pk
        if (($x instanceof int) || ($x instanceof PublisherRef)) {
            return new TagInfo("publisher", "pk", [ $x->id ]);
        }
        // name (lookup)
        if ($x instanceof str) {
            return new TagInfo("publisher", "name", [ $x ]);
        }
        // all
        if (is_null($x)) {
            return new TagInfo("publisher", "all");
        }
        return null;
    }
}

?>